File: MiniDOMSIP/Interfono/Interfono.cs
/* TODO ERROR: Skipped DefineDirectiveTrivia */
using System;
using System.Net.Sockets;
using DMD.XML;
using minidom.Win32;

namespace minidom.PBX
{
    [Serializable]
    public class Interfono : IDMDXMLSerializable, IComparable
    {
        private string m_UserName;
        private string m_Address;
        internal Office.Dispositivo Dev;
        internal Office.DeviceLog Log;
        internal InterfonoConnection Con;

        public Interfono()
        {
            m_UserName = "";
            m_Address = "";
        }

        public int UniqueID
        {
            get
            {
                return DBUtils.GetID(Dev);
            }
        }

        private void DoChanged(string pName, object newValue, object oldValue)
        {
        }

        void IDMDXMLSerializable.XMLSerialize(XMLWriter writer)
        {
            writer.WriteAttribute("UserName", m_UserName);
            writer.WriteAttribute("Address", m_Address);
        }

        void IDMDXMLSerializable.SetFieldInternal(string fieldName, object fieldValue)
        {
            switch (fieldName ?? "")
            {
                case "UserName":
                    {
                        m_UserName = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "Address":
                    {
                        m_Address = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }
            }
        }

        public string UserName
        {
            get
            {
                return m_UserName;
            }

            set
            {
                value = DMD.Strings.Trim(value);
                string oldValue = m_UserName;
                if ((oldValue ?? "") == (value ?? ""))
                    return;
                m_UserName = value;
                DoChanged("UserName", value, oldValue);
            }
        }

        public string Address
        {
            get
            {
                return m_Address;
            }

            set
            {
                value = DMD.Strings.Trim(value);
                string oldValue = m_Address;
                if ((oldValue ?? "") == (value ?? ""))
                    return;
                m_Address = value;
                DoChanged("Address", value, oldValue);
            }
        }

        public bool DisableMic = false;

        public override string ToString()
        {
            var ret = new System.Text.StringBuilder();
            ret.Append(UserName);
            ret.Append("   (");
            if (Dev is object)
            {
                ret.Append(Dev.Nome);
                ret.Append(" - ");
            }

            ret.Append(Address);
            ret.Append(")");
            return ret.ToString();
        }

        int IComparable.CompareTo(object obj)
        {
            return CompareTo((Interfono)obj);
        }

        public int CompareTo(Interfono obj)
        {
            return string.Compare(ToString(), obj.ToString(), true);
        }

        public bool IsConnected()
        {
            return Con is object && Con.client.Connected && Con.@params.Result == "DMDOK";
        }

        public InterfonoConnection Connect()
        {
            /* TODO ERROR: Skipped IfDirectiveTrivia *//* TODO ERROR: Skipped DisabledTextTrivia *//* TODO ERROR: Skipped EndIfDirectiveTrivia */
            /* TODO ERROR: Skipped IfDirectiveTrivia */
            PBX.Log.LogMessage("Interfono.Connect");
            PBX.Log.InternalLog("Interfono.Connect " + UserName + " (" + Address + ")");
            /* TODO ERROR: Skipped EndIfDirectiveTrivia */
            Con = new InterfonoConnection();
            Con.@params = new InterfonoParams();
            Con.@params.codec = Settings.WaveCodec;
            Con.@params.srcID = UniqueID;
            Con.@params.srcIP = Machine.GetIPAddress();
            Con.@params.srcPort = 10446;
            Con.@params.targetIP = Address;
            Con.@params.targetPort = 10446;

            /* TODO ERROR: Skipped IfDirectiveTrivia */
            PBX.Log.LogMessage("Interfono.Connect -> TcpClient.BeginConnect");
            /* TODO ERROR: Skipped EndIfDirectiveTrivia */
            Con.client = new TcpClient();
            IAsyncResult result;
            result = Con.client.BeginConnect(Address, 10445, null, null);
            bool success = result.AsyncWaitHandle.WaitOne(5000, true);
            if (Con.client.Connected)
            {
                Con.client.EndConnect(result);
            }
            else
            {
                // NOTE, MUST CLOSE THE SOCKET
                Con.client.Close();
                Con = null;
                throw new ApplicationException("Failed to connect server.");
            }

            /* TODO ERROR: Skipped IfDirectiveTrivia */
            PBX.Log.LogMessage("Interfono.Connect -> TcpClient.GetStream");
            /* TODO ERROR: Skipped EndIfDirectiveTrivia */
            Con.stream = Con.client.GetStream();
            /* TODO ERROR: Skipped IfDirectiveTrivia */
            PBX.Log.LogMessage("Interfono.Connect -> TcpClient.WriteParams");
            /* TODO ERROR: Skipped EndIfDirectiveTrivia */
            Sistema.BinarySerialize(Con.@params, Con.stream);
            /* TODO ERROR: Skipped IfDirectiveTrivia */
            PBX.Log.LogMessage("Interfono.Connect -> TcpClient.ReadParams");
            /* TODO ERROR: Skipped EndIfDirectiveTrivia */
            System.Threading.Thread.Sleep(1000);
            Con.@params = (InterfonoParams)Sistema.BinaryDeserialize(Con.stream);
            var res = DMD.Strings.Split(Con.@params.Result, "|");
            string errorcode = res[0];
            switch (errorcode ?? "")
            {
                case "DMDOK":
                    {
                        PBX.Log.InternalLog("Interfono.Connect (ok)");
                        break;
                    }

                case "DMDWARNING":
                    {
                        PBX.Log.LogMessage("Interfono.Connect -> Warning: " + res[1]);
                        // Toast.Show("Errore nell'handshake: " & res(1), 5000)
                        PBX.Log.InternalLog("Interfono.Connect (warning) " + res[1]);
                        break;
                    }

                default:
                    {

                        /* TODO ERROR: Skipped IfDirectiveTrivia */
                        PBX.Log.LogMessage("Interfono.Connect -> Error: " + res[1]);
                        PBX.Log.InternalLog("Interfono.Connect (error) " + res[1]);
                        /* TODO ERROR: Skipped EndIfDirectiveTrivia */
                        Con.stream.Close();
                        Con.client.Close();
                        Con = null;
                        throw new Exception("Errore nell'handshake: " + res[1]);
                        break;
                    }
            }


            /* TODO ERROR: Skipped IfDirectiveTrivia */
            PBX.Log.LogMessage("Interfono.Connect -> StartDevices");
            PBX.Log.InternalLog("Interfono.Connect (startdevices)");
            /* TODO ERROR: Skipped EndIfDirectiveTrivia */
            Con.StartDevices();

            /* TODO ERROR: Skipped IfDirectiveTrivia */
            PBX.Log.LogMessage("Interfono.Connect -> StartListening");
            PBX.Log.InternalLog("Interfono.Connect (startlistening)");
            /* TODO ERROR: Skipped EndIfDirectiveTrivia */
            Con.Peer = this;
            Con.StartListening();
            return Con;
            /* TODO ERROR: Skipped IfDirectiveTrivia *//* TODO ERROR: Skipped DisabledTextTrivia *//* TODO ERROR: Skipped EndIfDirectiveTrivia */
        }

        public InterfonoConnection Disconnect()
        {
            PBX.Log.InternalLog("Interfono.Disconnect " + UserName + " (" + Address + ")");
            var con = Con;
            con.StopDevices();
            con.SendDisconnectCommand();
            InternalDisconnect();
            Con = null;
            return con;
        }

        public bool IsAvailable()
        {
            if (!string.IsNullOrEmpty(Address) && Log is object)
            {
                var d = Log.EndDate;
                if (d.HasValue == false)
                    d = Log.StartDate;
                if (d.HasValue)
                {
                    if (DMD.DateUtils.DateDiff(DMD.DateTimeInterval.Minute, d.Value, DMD.DateUtils.Now()) <= 5L)
                        return true;
                }
            }

            return false;
        }

        internal void InternalDisconnect()
        {
            Con.StopListening();
            System.Threading.Thread.Sleep(100);
            Con.m_StreamThread.Abort();
            Con.m_StreamThread = null;
            if (Con.stream is object)
            {
                Con.stream.Dispose();
                Con.stream = null;
            }

            if (Con.client is object)
            {
                Con.client.Close();
                Con.client = null;
            }
        }
    }
}

File: MiniDOMSIP/Interfono/Interfono.vb
#Const VERBOSE = 1

Imports System.Drawing.Imaging
Imports minidom.Databases
Imports System.Drawing

Imports minidom.XML
Imports minidom.Office
Imports System.Net.Sockets
Imports System.IO
Imports minidom
Imports minidom.diallib
Imports minidom.Win32

<Serializable>
Public Class Interfono
    Implements minidom.XML.IDMDXMLSerializable, IComparable

    Private m_UserName As String
    Private m_Address As String
    Friend Dev As Dispositivo
    Friend Log As DeviceLog
    Friend Con As InterfonoConnection

    Public Sub New()
        Me.m_UserName = ""
        Me.m_Address = ""

    End Sub

    Public ReadOnly Property UniqueID As Integer
        Get
            Return GetID(Me.Dev)
        End Get
    End Property


    Private Sub DoChanged(ByVal pName As String, ByVal newValue As Object, ByVal oldValue As Object)

    End Sub

    Private Sub XMLSerialize(writer As XMLWriter) Implements IDMDXMLSerializable.XMLSerialize
        writer.WriteAttribute("UserName", Me.m_UserName)
        writer.WriteAttribute("Address", Me.m_Address)
    End Sub

    Private Sub SetFieldInternal(fieldName As String, fieldValue As Object) Implements IDMDXMLSerializable.SetFieldInternal
        Select Case fieldName
            Case "UserName" : Me.m_UserName = minidom.XML.Utils.Serializer.DeserializeString(fieldValue)
            Case "Address" : Me.m_Address = minidom.XML.Utils.Serializer.DeserializeString(fieldValue)
        End Select
    End Sub

    Public Property UserName As String
        Get
            Return Me.m_UserName
        End Get
        Set(value As String)
            value = Strings.Trim(value)
            Dim oldValue As String = Me.m_UserName
            If (oldValue = value) Then Return
            Me.m_UserName = value
            Me.DoChanged("UserName", value, oldValue)
        End Set
    End Property

    Public Property Address As String
        Get
            Return Me.m_Address
        End Get
        Set(value As String)
            value = Strings.Trim(value)
            Dim oldValue As String = Me.m_Address
            If (oldValue = value) Then Return
            Me.m_Address = value
            Me.DoChanged("Address", value, oldValue)
        End Set
    End Property

    Public DisableMic As Boolean = False

    Public Overrides Function ToString() As String
        Dim ret As New System.Text.StringBuilder
        ret.Append(Me.UserName)
        ret.Append("   (")
        If (Me.Dev IsNot Nothing) Then
            ret.Append(Me.Dev.Nome)
            ret.Append(" - ")
        End If
        ret.Append(Me.Address)
        ret.Append(")")
        Return ret.ToString
    End Function

    Private Function _CompareTo(obj As Object) As Integer Implements IComparable.CompareTo
        Return Me.CompareTo(CType(obj, Interfono))
    End Function

    Public Function CompareTo(ByVal obj As Interfono) As Integer
        Return String.Compare(Me.ToString, obj.ToString, True)
    End Function

    Public Function IsConnected() As Boolean
        Return Me.Con IsNot Nothing AndAlso Me.Con.client.Connected AndAlso Me.Con.params.Result = "DMDOK"
    End Function

    Public Function Connect() As InterfonoConnection
#If Not DEBUG Then
        Try
#End If

#If VERBOSE > 0 Then
        diallib.Log.LogMessage("Interfono.Connect")
        diallib.Log.InternalLog("Interfono.Connect " & Me.UserName & " (" & Me.Address & ")")
#End If

        Me.Con = New InterfonoConnection
        Me.Con.params = New InterfonoParams
        Me.Con.params.codec = Settings.WaveCodec
        Me.Con.params.srcID = Me.UniqueID
        Me.Con.params.srcIP = Machine.GetIPAddress
        Me.Con.params.srcPort = 10446
        Me.Con.params.targetIP = Me.Address
        Me.Con.params.targetPort = 10446

#If VERBOSE > 0 Then
        diallib.Log.LogMessage("Interfono.Connect -> TcpClient.BeginConnect")
#End If
        Me.Con.client = New TcpClient
        Dim result As IAsyncResult
        result = Me.Con.client.BeginConnect(Me.Address, 10445, Nothing, Nothing)
        Dim success As Boolean = result.AsyncWaitHandle.WaitOne(5000, True)
        If (Me.Con.client.Connected) Then
            Me.Con.client.EndConnect(result)
        Else
            ' NOTE, MUST CLOSE THE SOCKET
            Me.Con.client.Close()
            Me.Con = Nothing
            Throw New ApplicationException("Failed to connect server.")
        End If

#If VERBOSE > 0 Then
        diallib.Log.LogMessage("Interfono.Connect -> TcpClient.GetStream")
#End If
        Me.Con.stream = Me.Con.client.GetStream
#If VERBOSE > 0 Then
        diallib.Log.LogMessage("Interfono.Connect -> TcpClient.WriteParams")
#End If
        minidom.Sistema.BinarySerialize(Me.Con.params, Me.Con.stream)
#If VERBOSE > 0 Then
        diallib.Log.LogMessage("Interfono.Connect -> TcpClient.ReadParams")
#End If
        System.Threading.Thread.Sleep(1000)
        Me.Con.params = CType(minidom.Sistema.BinaryDeserialize(Me.Con.stream), InterfonoParams)
        Dim res As String() = Strings.Split(Me.Con.params.Result, "|")
        Dim errorcode As String = res(0)
        Select Case errorcode
            Case "DMDOK"
                diallib.Log.InternalLog("Interfono.Connect (ok)")
            Case "DMDWARNING"
                diallib.Log.LogMessage("Interfono.Connect -> Warning: " & res(1))
                diallib.Log.InternalLog("Interfono.Connect (warning) " & res(1))
                'Toast.Show("Errore nell'handshake: " & res(1), 5000)
            Case Else

#If VERBOSE > 0 Then
                diallib.Log.LogMessage("Interfono.Connect -> Error: " & res(1))
                diallib.Log.InternalLog("Interfono.Connect (error) " & res(1))
#End If
                Me.Con.stream.Close()
                Me.Con.client.Close()
                Me.Con = Nothing
                Throw New Exception("Errore nell'handshake: " & res(1))

        End Select


#If VERBOSE > 0 Then
        diallib.Log.LogMessage("Interfono.Connect -> StartDevices")
        diallib.Log.InternalLog("Interfono.Connect (startdevices)")
#End If
        Me.Con.StartDevices()

#If VERBOSE > 0 Then
        diallib.Log.LogMessage("Interfono.Connect -> StartListening")
        diallib.Log.InternalLog("Interfono.Connect (startlistening)")
#End If
        Me.Con.Peer = Me
        Me.Con.StartListening()

        Return Me.Con
#If Not DEBUG Then
        Catch ex As Exception
        If Me.Con IsNot Nothing Then
                If (Me.Con.InDev IsNot Nothing AndAlso Not Me.Con.InDev.IsDisposed) Then Me.Con.InDev.Dispose() : Me.Con.InDev = Nothing
                If (Me.Con.OutDev IsNot Nothing AndAlso Not Me.Con.OutDev.IsDisposed) Then Me.Con.OutDev.Dispose() : Me.Con.OutDev = Nothing
                If (Me.Con.stream IsNot Nothing) Then Me.Con.stream.Dispose() : Me.Con.stream = Nothing
                If (Me.Con.client IsNot Nothing) Then Me.Con.client.Close() : Me.Con.client = Nothing
            End If
            Me.Con = Nothing
            Throw
        End Try
#End If
    End Function

    Public Function Disconnect() As InterfonoConnection
        diallib.Log.InternalLog("Interfono.Disconnect " & Me.UserName & " (" & Me.Address & ")")
        Dim con As InterfonoConnection = Me.Con
        con.StopDevices()
        con.SendDisconnectCommand()
        Me.InternalDisconnect()
        Me.Con = Nothing
        Return con
    End Function

    Public Function IsAvailable() As Boolean
        If (Me.Address <> "" AndAlso Me.Log IsNot Nothing) Then
            Dim d As Date? = Me.Log.EndDate
            If (d.HasValue = False) Then d = Me.Log.StartDate
            If (d.HasValue) Then
                If (minidom.Sistema.DateUtils.DateDiff(DateInterval.Minute, d.Value, Now) <= 5) Then Return True
            End If
        End If
        Return False
    End Function

    Friend Sub InternalDisconnect()
        Me.Con.StopListening()
        System.Threading.Thread.Sleep(100)
        Me.Con.m_StreamThread.Abort()
        Me.Con.m_StreamThread = Nothing
        If (Me.Con.stream IsNot Nothing) Then Me.Con.stream.Dispose() : Me.Con.stream = Nothing
        If (Me.Con.client IsNot Nothing) Then Me.Con.client.Close() : Me.Con.client = Nothing
    End Sub
End Class
